### PR TITLE
Handle pointer items in ObjectList with ChildReconciler

### LINKED ISSUE #479: reconciler during runtime causes a panic
### Describe the bug

Go 1.22.0
reconciler-runtime 1.17.0

When Istio types are used with a ChildReconciler, there is a panic that happens here:

```
panic: interface conversion: interface {} is **v1beta1.DestinationRule, not *v1beta1.DestinationRule [recovered]
        panic: interface conversion: interface {} is **v1beta1.DestinationRule, not *v1beta1.DestinationRule

goroutine 96 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /Users/siva/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.0/pkg/internal/controller/controller.go:116 +0x1e5
panic({0x34aed40?, 0xc000633d70?})
        /Users/siva/go/go1.22.0/src/runtime/panic.go:770 +0x132
github.com/vmware-labs/reconciler-runtime/reconcilers.extractItems[...]({0x36dbf20, 0xc00020f570?})
        /Users/siva/go/pkg/mod/github.com/vmware-labs/reconciler-runtime@v0.17.0/reconcilers/child.go:391 +0x21d
github.com/vmware-labs/reconciler-runtime/reconcilers.(*ChildReconciler[...]).filterChildren(0x36e9060, 0xc000440c60, 0x36dbf20)
        /Users/siva/go/pkg/mod/github.com/vmware-labs/reconciler-runtime@v0.17.0/reconcilers/child.go:357 +0x45
github.com/vmware-labs/reconciler-runtime/reconcilers.(*ChildReconciler[...]).reconcile(0x36e9060, {0x36d4d18, 0xc000633830}, 0xc000440c60)
        /Users/siva/go/pkg/mod/github.com/vmware-labs/reconciler-runtime@v0.17.0/reconcilers/child.go:303 +0x2e5
github.com/vmware-labs/reconciler-runtime/reconcilers.(*ChildReconciler[...]).Reconcile(0x36e9060, {0x36d4d18, 0xc000633740}, 0xc000440c60)
        /Users/siva/go/pkg/mod/github.com/vmware-labs/reconciler-runtime@v0.17.0/reconcilers/child.go:258 +0x4f2
github.com/vmware-labs/reconciler-runtime/reconcilers.Sequence[...].Reconcile(0x0?, {0x36d4d18, 0xc000633650?}, 0xc000440c60?)
        /Users/siva/go/pkg/mod/github.com/vmware-labs/reconciler-runtime@v0.17.0/reconcilers/sequence.go:47 +0x206
github.com/vmware-labs/reconciler-runtime/reconcilers.(*ResourceReconciler[...]).reconcile(0x36e6f00, {0x36d4d18, 0xc000633650}, 0xc000440c60)
        /Users/siva/go/pkg/mod/github.com/vmware-labs/reconciler-runtime@v0.17.0/reconcilers/resource.go:273 +0x82
github.com/vmware-labs/reconciler-runtime/reconcilers.(*ResourceReconciler[...]).Reconcile(0x36e6f00, {0x36d4d18, 0xc0006333e0}, {{{0xc000136b58, 0x0?}, {0xc000138fd8?, 0x5?}}})
```

### Reproduction steps

Create a simple ChildController managing DestinationRules

```
func DestinationRule() reconcilers.SubReconciler[*v1alpha1.CustomCR] {
	return &reconcilers.ChildReconciler[*v1alpha1.CustomCR, *v1beta1.DestinationRule, *v1beta1.DestinationRuleList]{
		Name: "SimpleDestinationRule",
		DesiredChild: func(ctx context.Context, resource *v1alpha1.CustomCR) (*v1beta1.DestinationRule, error) {
			return &v1beta1.DestinationRule{}, nil
		},
		MergeBeforeUpdate: func(current, desired *v1beta1.DestinationRule) {},
		ReflectChildStatusOnParent: func(ctx context.Context, parent *v1alpha1.CustomCR, child *v1beta1.DestinationRule, err error) {
		},
	}
}
```

Running this reconciler will cause the panic

### Expected behavior

Running the reconciler should not cause a panic.

### Additional context

I believe this line using reflection is causing the issue:

https://github.com/vmware-labs/reconciler-runtime/blob/main/reconcilers/child.go#L391

Do note that the Istio Lists are a slice of pointers rather than the expected slice of types that are found in the k8s

https://github.com/istio/client-go/blob/d0d3864f139c210f15a0316565ede3da19beca31/pkg/apis/networking/v1beta1/types.gen.go#L75

### PR DESCRIPTION
**Pull request by scothis**
Canonically, Items in an ObjectList are a slice of structs. Some types, like Istio, use a slice of pointers intead. We now detect and handle both.

Resolves #479

### CODE REVIEW

**Comment by scothis** [comment]
@neowulf can you give this branch a try?

**Comment by mamachanko** [comment]
> @neowulf can you give this branch a try?

Because personally I can never recall the correct incantation, this may help:

```shell
go mod edit -replace github.com/vmware-labs/reconciler-runtime=github.com/scothis/reconciler-runtime@object-list-pointer
go mod tidy
```

File: reconcilers/util_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/vmware-labs/reconciler-runtime/internal/resources"
+	corev1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestExtractItems(t *testing.T) {
```
Comment (mamachanko) [inline reconcilers/util_test.go:17]: **[suggestion, non-blocking]:** Test the error case `"unknown type [...]"`.

**Review by mamachanko — APPROVED** [review]

**Comment by neowulf** [comment]
Thank you for the quick response! Yes, this patch works!

**Pull request merged**